addons/common/functions/fnc_switchPersistentLaser.sqf

```diff
@@ -17,51 +17,95 @@
+        _unit selectWeapon _weaponState;
+    };
+
+    private _fnc_turnOnLightLaser = {
+        params ["_unit", "_weaponState"];
+
+        if (currentWeapon _unit != _weaponState select 0) exitWith {};
+
+        action ["GunLightOn", _unit];
+        action ["IRLaserOn", _unit];
+
+        _unit selectWeapon _weaponState;
+    };
+
     [
+        // Wait for weapon in "ready to fire" direction
+        // If the unit changes weapons during the transition, let weapon EH handle it
         {
-            params ["_weapon", "_laserWasEnabled"];
-            private _laserEnabled = ACE_player isIRLaserOn _weapon || {ACE_player isFlashlightOn _weapon};
-            if (_laserEnabled && {_laserWasEnabled} || {!_laserEnabled && {!_laserWasEnabled}}) exitWith {};
-            private _weaponIndex = [ACE_player, _weapon] call FUNC(getWeaponIndex);
-            ACE_player setVariable [QGVAR(laserEnabled_) + str _weaponIndex, [nil, true] select _laserEnabled];
+            params ["_unit", "_weaponState"];
+
+            private _currentWeapon = currentWeapon _unit;
+
+            _currentWeapon != _weaponState select 0 ||
+            {0.01 > (getCameraViewDirection _unit) vectorDistance (_unit weaponDirection _currentWeapon)}
         },
-        [_weapon, ACE_player isIRLaserOn _weapon || {ACE_player isFlashlightOn _weapon}]
-    ] call CBA_fnc_execNextFrame;
-    false
-}] call CBA_fnc_addDisplayHandler;
-
-private _laserEH = {
-    if (sunOrMoon == 1) exitWith {};
-    params ["_player"];
-    private _weaponIndex = [_player, currentWeapon _player] call FUNC(getWeaponIndex);
-    if (
-        !(_player getVariable [QGVAR(laserEnabled_) + str _weaponIndex, false])
-        || {_weaponIndex > 0 && {"" != primaryWeapon _player}} // Arma switches to primary weapon if exists
-        || {!(_player call CBA_fnc_canUseWeapon)} // ignore in vehicle except FFV
-    ) exitWith {};
-    [
-        // wait for weapon in "ready to fire" direction
-        {0.01 > getCameraViewDirection _this vectorDistance (_this weaponDirection currentWeapon _this)},
-        {{_this action [_x, _this]} forEach ["GunLightOn", "IRLaserOn"]},
-        _player,
+        _fnc_turnOnLightLaser,
+        [_unit, _weaponState],
         3,
-        {{_this action [_x, _this]} forEach ["GunLightOn", "IRLaserOn"]}
+        _fnc_turnOnLightLaser
     ] call CBA_fnc_waitUntilAndExecute;
 };
 
-GVAR(laserLoadoutEH) = ["loadout", _laserEH] call CBA_fnc_addPlayerEventHandler;
-GVAR(laserTurretEH) = ["turret", _laserEH] call CBA_fnc_addPlayerEventHandler;
-GVAR(laserVehicleEH) = ["vehicle", _laserEH] call CBA_fnc_addPlayerEventHandler;
-GVAR(laserWeaponEH) = ["weapon", _laserEH] call CBA_fnc_addPlayerEventHandler;
+// Dropping weapons turns off lights/lasers
+GVAR(laserLoadoutEH) = ["loadout", _fnc_lightLaserEH] call CBA_fnc_addPlayerEventHandler;
```

Comment:
small oddity of `loadout`:
open inventory and drop a bandage on the ground it causes my laser to turn off

Reply:
I'm guessing the laser turns off temporarily? Or does it turn off and not turn on again?

Reply:
It's because dropping weapons causes lights/lasers to turn off. E.g: I have my primary weapon selected with visible laser on and I drop my launchers -> laser turns off.

Reply:
Personally I'd like to remove the alignment code, even if it's less realistic. Removing it would make the code simpler.
You can already turn on lasers/lights before your weapon is ready to fire, so imo it would be fine to remove the current behaviour.

Reply:
I went ahead and removed the alignment code. I also added some inventory changed EH filter, so that it should trigger when weapons are dropped/picked up.